system/mrm_handler/src/mrm_handler/mrm_handler_core.cpp

```diff
@@ -0,0 +1,587 @@
+      "waiting for mrm comfortable stop to become available...");
+    return false;
+  }
+
+  if (
+    mrm_emergency_stop_status_->state == tier4_system_msgs::msg::MrmBehaviorStatus::NOT_AVAILABLE) {
+    RCLCPP_INFO_THROTTLE(
+      this->get_logger(), *this->get_clock(), std::chrono::milliseconds(5000).count(),
+      "waiting for mrm emergency stop to become available...");
+    return false;
+  }
+
+  return true;
+}
+
+void MrmHandler::onTimer()
+{
+  if (!isDataReady()) {
+    return;
+  }
+  const bool is_operation_mode_availability_timeout =
+    (this->now() - stamp_operation_mode_availability_).seconds() >
+    param_.timeout_operation_mode_availability;
+  if (is_operation_mode_availability_timeout) {
+    RCLCPP_WARN_THROTTLE(
+      this->get_logger(), *this->get_clock(), std::chrono::milliseconds(1000).count(),
+      "heartbeat operation_mode_availability is timeout");
+    mrm_state_.state = autoware_adapi_v1_msgs::msg::MrmState::MRM_OPERATING;
+    publishControlCommands();
+    return;
+  }
+
+  // Update Emergency State
+  updateMrmState();
+
+  // Publish control commands
+  publishControlCommands();
+  operateMrm();
+  publishMrmState();
+}
+
+void MrmHandler::transitionTo(const int new_state)
+{
+  using autoware_adapi_v1_msgs::msg::MrmState;
+
+  const auto state2string = [](const int state) {
+    if (state == MrmState::NORMAL) {
+      return "NORMAL";
+    }
+    if (state == MrmState::MRM_OPERATING) {
+      return "MRM_OPERATING";
+    }
+    if (state == MrmState::MRM_SUCCEEDED) {
+      return "MRM_SUCCEEDED";
+    }
+    if (state == MrmState::MRM_FAILED) {
+      return "MRM_FAILED";
+    }
+
+    const auto msg = "invalid state: " + std::to_string(state);
+    throw std::runtime_error(msg);
+  };
+
+  RCLCPP_INFO(
+    this->get_logger(), "MRM State changed: %s -> %s", state2string(mrm_state_.state),
+    state2string(new_state));
+
+  mrm_state_.state = new_state;
+}
+
+void MrmHandler::updateMrmState()
+{
+  using autoware_adapi_v1_msgs::msg::MrmState;
+  using autoware_auto_vehicle_msgs::msg::ControlModeReport;
+
+  // Check emergency
+  const bool is_emergency = isEmergency();
+
+  // Get mode
+  const bool is_auto_mode = control_mode_->mode == ControlModeReport::AUTONOMOUS;
+  const bool is_takeover_done = control_mode_->mode == ControlModeReport::MANUAL;
+
+  // State Machine
+  if (mrm_state_.state == MrmState::NORMAL) {
+    // NORMAL
+    if (is_auto_mode && is_emergency) {
+      if (param_.use_takeover_request) {
+        takeover_requested_time_ = this->get_clock()->now();
+        is_takeover_request_ = true;
+        return;
+      } else {
+        transitionTo(MrmState::MRM_OPERATING);
+        return;
+      }
+    }
+  } else {
+    // Emergency
+    // Send recovery events if "not emergency" or "takeover done"
+    if (!is_emergency) {
+      transitionTo(MrmState::NORMAL);
+      return;
+    }
+    // TODO(Kenji Miyake): Check if human can safely override, for example using DSM
```

Comment:
It looks better to change `TODO(Kenji Miyake)` to someone else

Reply:
I will temporarily change `TODO(Kenji Miyake)` in charge to me.
I would like to discuss later the necessity of the modification and the content of the modification.